src/ttlv/fast_scan.rs

```diff
@@ -0,0 +1,1172 @@
+            const LENGTH_MASK: u32 = !(usize::MAX as u32);
+
+            // Generate a mask for the whole header.
+            //
+            // We always check the top 4 bytes (tag and type).
+            const MASK: u64 = u64::MAX << 32 | LENGTH_MASK as u64;
+
+            // Prepare an expected value as a native-endian integer.
+            //
+            // The length field is set to all-zeros; we will check that 'header'
+            // contains the same bits, but only for 'LENGTH_MASK'.
+            //
+            // The compiler should be able to fold this computation into a
+            // simple hard-coded 64-bit constant.
+            let expected = (tag.with_type(r#type).value() as u64) << 32;
+
+            // Load the header into the right endianness, as we need to read out
+            // its low bits for the length field.  Note that it and 'expected'
+            // have matching byte layouts.
+            let actual = u64::from_be_bytes(*header);
+
+            // Make the failure word non-zero if at least one bit in the actual
+            // header (of the tag, type, or high part of length) did not match.
+            let mismatch = (actual ^ expected) & MASK;
+
+            // We have already checked that the length fits in a 'usize', so we
+            // can freely truncate the value now.
+            (mismatch, actual as u32 as usize)
+        }
+    }
+}
+
+impl<'a> FastScanner<'a> {
+    /// Scan an expected structure.
+    ///
+    /// The returned sub-scanner can be used to scan the fields of the
+    /// `struct`; `self` will independently continue from the end of the
+    /// `struct`.
+    #[must_use = "Use the returned scanner to parse the structure, or use `FastScanner::skip_struct()` to ignore its contents"]
+    #[inline]
+    pub fn scan_struct(&mut self, tag: Tag) -> Result<Self, FastScanError> {
+        let [header, rest @ ..] = self.remaining else {
+            return Err(FastScanError::assert());
+        };
+        let length = self.check_tt_8l(header, tag, Type::Structure);
+        let (body, rest) = rest.split_at_checked(length).ok_or(FastScanError::assert())?;
+        self.remaining = rest;
+
+        Ok(Self::from_blocks(body))
+    }
+
+    /// Scan an expected structure with a pre-determined length.
+    ///
+    /// The returned sub-scanner can be used scan the fields of the `struct`;
+    /// `self` will independently continue from the end of the `struct`.  If
+    /// the length encoded in the input is different from the pre-determined
+    /// length, the error will be delayed and both scanners will return
+    /// garbage.
+    #[must_use = "Use the returned scanner to parse the structure, or use `FastScanner::skip_fixed_struct()` to ignore its contents"]
+    #[inline]
+    pub fn scan_fixed_struct(&mut self, tag: Tag, length: u32) -> Result<Self, FastScanError> {
+        debug_assert!(length.is_multiple_of(8));
+        if self.remaining.len() < 1 + length as usize / 8 {
```

Comment:
Would it be beneficial to store the result of the `length as usize / 8` calculation as it is done again below?